types/underscore/index.d.ts

```diff
@@ -115,6 +117,38 @@ declare module _ {
         : V extends Dictionary<infer T> ? T
         : never;
 
+    // given a union type like { a: string } | { b: number } | undefined, generates a union of all the keys of all the types in the union
+    type KeysOfUnion<T> = T extends any ? keyof T : never;
+
+    // given a union type like { a: string } | { a: number } | { b: boolean } and a property name,
+    // generates a union of all the types implemented by that property name and adds undefined
+    // if any type in the union doesn't implement the property
+    // e.g. TypesOfUnionProperty<{ a: string } | { a: number } | { b: boolean }, 'a'> => string | number | undefined
+    type TypesOfUnionProperty<T, K extends KeysOfUnion<T>> =
+        // iterate over each type T in a type union
+        T extends any
+        // check whether the property is implemented by T, using Extract to get around K possibly not being in keyof T
+        ? T[Extract<keyof T, K>] extends never
+        // if T does not implement property K, add undefined to the resulting type union
+        ? undefined
+        // if T does implement property K, add type T[K] to the resulting type union
+        : T[Extract<keyof T, K>]
+        // this should never actually be evaluated since all types extend any
+        : never;
+
+    // '& object' prevents strings from being matched by list checks
```

Comment:
Interesting point about `String`, I'll have to try that.

I feel like I ran into issues where I was ending up with `string` instead of `string[]` somehow, but that might have been back when the name of `ShallowFlattenedList` made sense and I was actually getting out an array type rather than an item type, so I'll see if this can be removed.

Reply:
I removed `& object` from `ListItemOrSelf` since it is not useful there but left it on `DeepestListItem`. Interestingly `TypeOfList<String>` is `string`, which can help keep `String` from recursing its way all the way to `any`, but it does go one extra level compared to other types. That being said, in the worst case folks will end up at `any[]`, and I think it serves them right for using `String` in the first place considering TS documentation itself warns not to do that 😛 

![image](https://user-images.githubusercontent.com/8692126/86430072-2637b700-bca6-11ea-97b5-a4b4f830f918.png)

Reply:
Derp, and just after I posted that I noticed that that's ESLint complaining rather than TS, oh well.